autoemulate/hyperparam_searching.py

```diff
@@ -0,0 +1,175 @@
+    elif search_type == "bayes":
+        searcher = BayesSearchCV(
+            model,
+            param_grid,
+            n_iter=niter,
+            cv=cv,
+            n_jobs=n_jobs,
+            refit=True,
+        )
+    elif search_type == "grid":
+        raise NotImplementedError("Grid search not available yet.")
+
+    # run hyperparameter search
+    try:
+        searcher.fit(X, y)
+    except Exception as e:
+        logger.info(
+            f"Failed to perform hyperparameter search on {get_model_name(model)}"
+        )
+        logger.info(e)
+    logger.info(f"Best parameters for {model_name}: {searcher.best_params_}")
+
+    return searcher.best_estimator_
+
+
+def process_param_grid(model, search_type, param_grid):
+    """Process parameter grid for hyperparameter search.
+    Gets the parameter grid for the model and adjusts it to include prefixes
+    for pipelines / multioutput estimators.
+
+    Parameters
+    ----------
+    model : model instance to do hyperparameter search for.
+    search_type : str, default="random"
+        Type of search to perform. Can be "random" or "bayes", "grid" not yet implemented.
+    param_grid : dict, default=None
+        Dictionary with parameters names (string) as keys and lists of
+        parameter settings to try as values, or a list of such dictionaries,
+        in which case the grids spanned by each dictionary in the list are
+        explored. This enables searching over any sequence of parameter
+        settings.
+
+    Returns
+    -------
+    param_grid : dict
+        Adjusted parameter grid.
+    """
+    # get param_grid if not provided
+    if param_grid is None:
+        param_grid = get_model_param_grid(model, search_type)
+    else:
+        param_grid = check_param_grid(param_grid, model)
+    # include prefixes for pipelines / multioutput estimators
+    param_grid = adjust_param_grid(model, param_grid)
+    return param_grid
+
+
+def check_param_grid(param_grid, model):
+    """Checks that the parameter grid is valid.
+
+    Parameters
+    ----------
+    param_grid : dict, default=None
+        Dictionary with parameters names (string) as keys and lists of
+        parameter settings to try as values, or a list of such dictionaries,
+        in which case the grids spanned by each dictionary in the list are
+        explored. This enables searching over any sequence of parameter
+        settings. Parameters names should be prefixed with "model__" to indicate that
+        they are parameters of the model.
+    model : sklearn.pipeline.Pipeline
+        Model to be optimized.
+
+    Returns
+    -------
+    param_grid : dict
+    """
+    if type(param_grid) != dict:
+        raise TypeError("param_grid must be a dictionary")
+    for key, value in param_grid.items():
+        if type(key) != str:
+            raise TypeError("param_grid keys must be strings")
+        if type(value) != list:
+            raise TypeError("param_grid values must be lists")
```

Comment:
Is there a logic behind how you raise some errors directly and log others? It looked like logging seems to focus on the fitting errors, while other more basic errors are flagged directly without being logged. I think this is fine if that is what you intended, but I wanted to be sure this is what you intended!

Reply:
Good catch. Haven't carefully thought about logging yet as I'm a bit unsure what would be a good strategy here. But this is an issue too #36 .